mysqlnd_azure.c

```diff
@@ -410,19 +436,32 @@ MYSQLND_METHOD(mysqlnd_azure_data, connect)(MYSQLND_CONN_DATA ** pconn,
                 port = ui_redirect_port;
                 transport = redirect_transport;
 
-                DBG_ENTER("[redirect]: mysql redirect handshake succeeded.");
+            } else { //redirect failed. if REDIRECT_ON, also abort the original conn, if REDIRECT_PREFERRED, use original connection
+                DBG_ENTER("[redirect]: mysql redirect handshake fails");
 
-            }
-            else { //redirect failed. use original connection information
+                if (MYSQLND_AZURE_G(enableRedirect) == REDIRECT_PREFERRED) {
+                    //free resource and use original connection
+                    redirect_conn->m->dtor(redirect_conn);
+                    if (redirect_transport.s) {
+                        mnd_sprintf_free(redirect_transport.s);
+                        redirect_transport.s = NULL;
+                    }
+                    goto after_conn;
+
+                } else { //REDIRECT_ON, free original connect, and use redirect_conn to handle error
+                    conn->m->send_close(conn);
```

Comment:
you still need to free redirect_transport here? if so, please try to merge free code in above if/else clause

Reply:
redirect_transport  is assigned to transport when we success handshake

Reply:
Yes, it need be fre here. I was think handling it in goto err section, but it is with different name. Thanks.

Reply:
the success condition is handled before. So it will be no problem